scripts/entity/components/movement_components/player_movement_component.gd

```diff
@@ -1,61 +1,134 @@
 ## The time it takes to fully decelerate from horizontal input in seconds.
 @export_range(0, 10, 0.1, "or_greater") var DECELERATION_TIME: float = 0.2
 
+## Moth Jump variables
+@export var MOTH_JUMP_HANGTIME: float = 1.0
+@export var MOTH_JUMP_DIP_AMOUNT: float = 18.0
+@export var MOTH_JUMP_FLUTTER_DURATION: float = 0.35
+
 ## The current input direction.
 var _current_input_direction: float = 0
 ## How fast the character decelerates after loss of input.
 var _horizontal_deceleration: float = MAX_SPEED / DECELERATION_TIME
 
-# Called when the node enters the scene tree for the first time.
+## Moth Jump state variables
+var _is_moth_jumping: bool = false
+var _can_moth_jump: bool = true
+var _is_fluttering: bool = false
+var _moth_jump_timer: float = 0.0
+var _moth_jump_flutter_timer: float = 0.0
+var _moth_jump_initial_height: float = 0.0
+var _is_jump_button_held: bool = false
+
+
 func _ready() -> void:
-	#do nothing
+	JUMP_VELOCITY *= -1
 	pass
 
-
 ## Applies movement to the character.
 func _physics_process(delta: float) -> void:
 	super._physics_process(delta)
-	apply_gravity(delta)	
+	apply_gravity(delta)
 	apply_horizontal_movement(delta)
 	move_and_slide()
 
-
 func on_movement_input(horizontal_vector2: float) -> void:
 	_current_input_direction = horizontal_vector2
 
-
 func on_jump_input_started() -> void:
 	if is_on_floor():
+		_can_moth_jump = true
 		velocity.y = JUMP_VELOCITY
 		on_jump_started.emit()
-
+	elif not _is_moth_jumping and _can_moth_jump:
+		# Start Moth Jump
+		_init_moth_jump()
 
 func on_jump_input() -> void:
-	# do nothing
-	pass
-
+	_is_jump_button_held = true
 
 func on_jump_input_cancelled() -> void:
-	# do nothing
-	pass
-
+	_is_jump_button_held = false
 
 ## Moves the entity based on the current input direction.
 func apply_horizontal_movement(delta: float) -> void:
- 	# Direction button active
-	if _current_input_direction:
+	if _current_input_direction != 0:
 		# Accelerate towards the target speed
 		velocity.x = move_toward(velocity.x, _current_input_direction * MAX_SPEED, ACCELERATION * delta)
-	 # Direction button inactive
 	else:
 		# Decelerate towards zero speed
 		velocity.x = move_toward(velocity.x, 0, _horizontal_deceleration * delta)
+
+func apply_gravity(delta: float) -> void:
+	if _is_moth_jumping:
+		_can_moth_jump = false
+		if _is_fluttering:
+			
+			_apply_moth_jump_dip(delta)
+			
+			_moth_jump_flutter_timer += delta
+			if _moth_jump_flutter_timer >= MOTH_JUMP_FLUTTER_DURATION or !_is_jump_button_held:
+				_is_fluttering = false
+				velocity.y = 0.0
+				on_moth_jump_ended.emit()
+
+		else:
+			if _is_jump_button_held and _moth_jump_timer < MOTH_JUMP_HANGTIME:
+				# Maintain the current height
+				velocity.y = 0.0
+				_moth_jump_timer += delta
+
+			else:
+				# End Moth Jump
+				_is_moth_jumping = false
+				on_moth_jump_ended.emit()
+				super.apply_gravity(delta)  # Resume normal gravity
+
+	else:
+		# Apply normal gravity
+		super.apply_gravity(delta)
```

Comment:
Small organization concern, but it'd be better to early exit and have less nesting. Video example posting in programming section of discord (https://www.youtube.com/watch?v=CFRhGnuXG-4&pp=ygUMbmV2ZXIgbmVzdGVy)

Reply:
Nested less

Reply:
Id just be concerned mostly about readability with it